### PR TITLE
Full color conversions and luminance management

### PR DESCRIPTION
**Pull request by pmariac**
Added all converters in Color interface (from any colorspace to any other : RGB, HCL, HSL, LAB)
Internalize the converter functions from the colorConversions class

Add a "perceived lightness" management with luminance() and contrast() according to W3C standards.

Add a "hue" modifier that relies on HCL colorspace (HCL is design to respect human perception so same step in hue from 2 different colors should produce the same "delta perception".

Add some tests.

### CODE REVIEW

File: color/d2v-color-common/src/main/kotlin/io/data2viz/color/Color.kt [hunk 1]
```diff
@@ -19,15 +21,59 @@ interface Color : ColorOrGradient {
...
+     * A minimum contrast of 4.5:1 is recommended to ensure that text is still readable against a background color.
+     * According to the WCAG definition: https://www.w3.org/TR/2008/REC-WCAG20-20081211/#contrast-ratiodef
+     */
+    fun contrast(other:Color):Double
+
+    /**
+     * Return `true` if 2 colors have a contrast of 4.5 or more.
+     * According to the WCAG definition: https://www.w3.org/TR/WCAG20-TECHS/G18.html
+     */
+    fun isContrastOK(other:Color):Boolean
```
Comment (gzoritchak) [inline color/d2v-color-common/src/main/kotlin/io/data2viz/color/Color.kt]: The referenced article is very precise about its target: we are talking about the contrast needed between a text and its background. The name of the function should represent this purpose.

> this technique describes the minimum contrast ratio for text that is less than 18 point (if not bold) and less than 14 point (if bold). For Success Criterion 1.4.5, this technique relaxes the 7:1 contrast ratio requirement for text that is at least 18 point (if not bold) or at least 14 point (if bold).

File: color/d2v-color-common/src/main/kotlin/io/data2viz/color/HslColor.kt [hunk 2]
```diff
@@ -26,12 +26,23 @@ internal constructor(hue: Angle, saturation: Double, lightness: Double, a: Doubl
...
+    override fun toLab(): LabColor = toRgb().toLab()
+    override fun toHcl(): HclColor = toRgb().toHcl()
+    override fun toHsl(): HslColor = this
+
     override fun brighten(strength: Double): Color = toRgb().brighten(strength)
     override fun darken(strength: Double): Color = toRgb().darken(strength)
     override fun saturate(strength: Double): Color = toRgb().saturate(strength)
     override fun desaturate(strength: Double): Color = toRgb().desaturate(strength)
     override fun withAlpha(alpha: Double) = Colors.hsl(h, s, l, alpha)
+//    override fun withLuminance(luminance: Percent) = toLab().withLuminance(luminance)
```
Comment (gzoritchak) [inline color/d2v-color-common/src/main/kotlin/io/data2viz/color/HslColor.kt]: remove

File: color/d2v-color-common/src/main/kotlin/io/data2viz/color/HslColor.kt [hunk 3]
```diff
@@ -26,12 +26,23 @@ internal constructor(hue: Angle, saturation: Double, lightness: Double, a: Doubl
     override val b = toRgb().b
     override val rgbHex: String = toRgb().rgbHex
 
+    override fun luminance() = toRgb().luminance()
+    override fun contrast(other:Color) = toRgb().contrast(other)
+    override fun isContrastOK(other: Color) = toRgb().isContrastOK(other)
+//    override val hue: Angle = toHcl().hue
```
Comment (gzoritchak) [inline color/d2v-color-common/src/main/kotlin/io/data2viz/color/HslColor.kt]: to remove.

File: color/d2v-color-common/src/main/kotlin/io/data2viz/color/HclColor.kt [hunk 4]
```diff
@@ -26,12 +30,23 @@ internal constructor(val h: Angle, val c: Double, luminance: Double, a: Double =
...
+    override fun toLab(): LabColor = toLaba()
+    override fun toHcl(): HclColor = this
+    override fun toHsl(): HslColor = toLab().toHsl()
+
     override fun brighten(strength: Double): Color = Colors.hcl(h, c, (l + (Kn * strength)), alpha)
     override fun darken(strength: Double): Color = Colors.hcl(h, c, (l - (Kn * strength)), alpha)
     override fun saturate(strength: Double): Color = Colors.hcl(h, max(.0, (c + (Kn * strength))), l, alpha)
     override fun desaturate(strength: Double): Color = Colors.hcl(h, max(.0, (c - (Kn * strength))), l, alpha)
     override fun withAlpha(alpha: Double) = Colors.hcl(h, c, l, alpha)
+//    override fun withLuminance(luminance: Percent) = Colors.hcl(h, c, luminance.value*100, alpha)
```
Comment (gzoritchak) [inline color/d2v-color-common/src/main/kotlin/io/data2viz/color/HclColor.kt]: to remove.

File: color/d2v-color-common/src/main/kotlin/io/data2viz/color/ColorConversions.kt [hunk 5]
```diff
@@ -1,6 +1,7 @@
 package io.data2viz.color
 
 import io.data2viz.math.Angle
+import io.data2viz.math.Percent
```
Comment (gzoritchak) [inline color/d2v-color-common/src/main/kotlin/io/data2viz/color/ColorConversions.kt]: use star import.

File: color/d2v-color-common/src/main/kotlin/io/data2viz/color/LabColor.kt [hunk 6]
```diff
@@ -25,12 +27,23 @@ internal constructor(lightness: Double, aComponent: Double, bComponent: Double,
...
+    override fun toHsl(): HslColor = toRgb().toHsl()
+
     override fun brighten(strength: Double): Color = Colors.lab((labL + (Kn * strength)), labA, labB, alpha)
     override fun darken(strength: Double): Color = Colors.lab((labL - (Kn * strength)), labA, labB, alpha)
-    override fun saturate(strength: Double): Color = toHcla().saturate(strength)
-    override fun desaturate(strength: Double): Color = toHcla().desaturate(strength)
+    override fun saturate(strength: Double): Color = toHcl().saturate(strength)
+    override fun desaturate(strength: Double): Color = toHcl().desaturate(strength)
     override fun withAlpha(alpha: Double) = Colors.lab(labL, labA, labB, alpha)
+//    override fun withLuminance(luminance: Percent) = Colors.lab(luminance.value * 100, labA, labB, alpha)
```
Comment (gzoritchak) [inline color/d2v-color-common/src/main/kotlin/io/data2viz/color/LabColor.kt]: to remove.

File: color/d2v-color-common/src/main/kotlin/io/data2viz/color/Color.kt (hunk 1 above)
Comment (pmariac) [reply color/d2v-color-common/src/main/kotlin/io/data2viz/color/Color.kt]: I remove it, but "luminance" and "contrast" will remains and will certainly be used in charts.kt (in a more precise way including font size etc...)

File: color/d2v-color-common/src/main/kotlin/io/data2viz/color/HclColor.kt [hunk 7]
```diff
@@ -1,22 +1,26 @@
...
  * Create a color in the HCL color space (CIELCH)
  *
+ * This Colorspace is designed to accord with human perception of color.
+ * HCL has been adopted by information visualization practitioners to present data without the bias implicit in
+ * using varying saturation.
+ *
  * @param h hue: Angle in degree
  * @param c chroma: Float, the upper bound for chroma depends on hue and luminance (typically in 0..230)
- * @param luminance: Float a value in the range [0,100] giving the luminance of the colour (in percent)
+ * @param lightness: Float a value in the range [0,100] giving the luminance of the colour (in percent)
```
Comment (gzoritchak) [inline color/d2v-color-common/src/main/kotlin/io/data2viz/color/HclColor.kt]: " giving the luminance " shouldn't it be lightness?

**Pull request merged**